drivers/gpio/gpio_altera_nios2.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file Driver for the Altera Nios-II PIO Core.
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <gpio.h>
+
+#include "gpio_utils.h"
+#include "altera_avalon_pio_regs.h"
+
+typedef void (*setup_func_t)(struct device *dev);
```

Comment:
I'm pretty confused on why we need a function pointer indirection layer here.
In what cases in config_func NULL or some other function besides gpio_nios2_config_oput_port?
setup_func is not used at all...

I could maybe understand this if these were specified and instantiated outside of this C file, but since this is all in the same C file I do not see the point...

Reply:
Nios2 PIO core can configure in different ways output only, input only and bidirectional.
For output only case we don't need to configure interrupt as PIO will not generate any. But if the PIO is configured as input then we need to set up the interrupt-related routines and flags accordingly.
These kinds of activities should be done from setup_ func(). In case of output only port we don't need to do anything as HW is already configured properly.

we will pass setup_func() primarily for input only or bidirectional ports.

Reply:
I can remove the setup_func indirection and directly pass the init function to the DEVICE_AND_API_INIT(). This should avoid the confusion :)